plugins/oidc/access.go

```diff
@@ -10,118 +10,207 @@ import (
+//	err (error): the error, if any occurred during the process.
+func discoverEndpoints(discoveryEndpoint string) (userinfoEndpoint string, jwksEndpoint string, err error) {
+	// cleanup the discovery endpoint address and remove possible whitespaces
+	discoveryEndpoint = strings.TrimSpace(discoveryEndpoint)
+	if discoveryEndpoint == "" {
+		return "", "", errEmptyEndpoint
+	}
+
+	// now request the openid connect configuration endpoint
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
+}
+
+// extractBearerToken is a function that takes an authorization header value as
+// input and extracts the bearer token from it.
+// It uses a regular expression to match the header value against a predefined
+// pattern.
+// If a match is found and there are two capture groups in the match result, the
+// function returns the second capture group, which represents the bearer token.
+// If no match is found or there are not exactly two capture groups, an empty
+// string is returned.
+//
+// Example usage:
+//
+//	token := extractBearerToken("Bearer abcxyz123")
+//	// token = "abcxyz123"
+//
+// Parameters:
+//   - authorizationHeaderValue: The value of the authorization header.
+//
+// Returns:
+//
+//	The extracted bearer token, or an empty string if no valid bearer token is found.
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
+	matches := regex.FindStringSubmatch(authorizationHeaderValue)
+	if len(matches) == 3 {
+		return matches[2]
+	}
+	return ""
+}
+
+// Access is executed every time the kong gateway receives a request. Since ther
 // plugin may be restarted at any moment
 func (c *Configuration) Access(kong *pdk.PDK) {
-	// access the request and get the authorization header
-	request := kong.Request
+	// create a shortcut for the logger
 	logger := kong.Log
-	logger.Info("authenticating new request")
-	logger.Debug("extracting authorization header and its values")
-	authorizationHeader, _ := request.GetHeader("Authorization")
-	// now trim away any blank values at the start or end of the string
-	authorizationHeader = strings.TrimSpace(authorizationHeader)
-	// now get the authorization method and token from the header
-	matches := headerRegex.FindStringSubmatch(authorizationHeader)
-	if len(matches) != 3 {
-		// since the matches array needs to contain three entries, the value is
-		// invalid and therefore the user is not authorized
+
+	userinfoEndpoint, jwksEndpoint, err := discoverEndpoints(c.DiscoveryUri)
```

Comment:
The endpoint discovery always runs a GET request which takes some time. Maybe this is a bit too rigorous.

Reply:
True, however this is the only way a configuration change during the plugin runtime can be caught since the plugin does not need to be restarted if the plugin configuration has been changed and caching these values may result in an validation error further down in the code